src/main/java/com/gargoylesoftware/htmlunit/javascript/host/dom/ShadowRoot.java

```diff
@@ -14,27 +14,228 @@
+     */
+    @JsxGetter({CHROME, EDGE, FF, FF_ESR})
+    public Element getActiveElement() {
+        
+        return null;
+    }
+
+    /**
+     * Returns the value of the {@code adoptedStyleSheets} property.
+     * @see <a href="https://developer.mozilla.org/en-US/docs/Web/API/Document/adoptedStyleSheets">MDN documentation</a>
+     * @return the value of the {@code adoptedStyleSheets} property
+     */
+    @JsxGetter({CHROME, EDGE, FF, FF_ESR})
+    public CSSStyleSheet[] getAdoptedStyleSheets() {
+        return adoptedStyleSheets_;
+    }
+    
+    /**
+     * Set the array of style sheets to be used by the shadow DOM subtree.
+     * @param styleSheets array of {@link CSSStyleSheet} objects to be adopted
+     * @throws DOMException <i>NotAllowedError</i> if a specified style sheet is not owned by the current document
+     */
+    @JsxSetter({CHROME, EDGE, FF, FF_ESR})
+    public void setAdoptedStyleSheets(final CSSStyleSheet... styleSheets) throws DOMException {
+        Object ownerDocument = getOwnerDocument();
+        CSSStyleSheet[] adoptedStyleSheets = new CSSStyleSheet[styleSheets.length];
+        // iterate over style sheet adoptees
+        for (int i = 0; i < styleSheets.length; i++) {
+            HTMLElement ownerNode = styleSheets[i].getOwnerNode();
+            // if style sheet owned
+            if (ownerNode != null) {
+                // if style sheet not owned by shadow root document
+                if (ownerNode.getOwnerDocument() != ownerDocument) {
+                    throw Context.reportRuntimeError("NotAllowedError: Style sheet was not constructed in the current document.");
+                }
+            }
+            // save reference to style sheet
+            adoptedStyleSheets[i] = styleSheets[i];
+        }
+        adoptedStyleSheets_ = adoptedStyleSheets;
+    }
+    
+    @JsxGetter({CHROME, EDGE, FF, FF_ESR})
+    public boolean getDelegatesFocus() {
+        return delegatesFocus_;
+    }
+    
+    @Override
+    public void setEventHandler(final String eventType, final Object value) {
+        super.setEventHandler(eventType, value);
+    }
+    
+    @JsxGetter({CHROME, EDGE, FF, FF_ESR})
+    public Function getOnslotchange() {
+        return getEventHandler(Event.TYPE_SLOT_CHANGE);
+    }
+    
+    @JsxSetter({CHROME, EDGE, FF, FF_ESR})
+    public void setOnslotchange(final EventHandler slotchange) {
+        setEventHandler(Event.TYPE_SLOT_CHANGE, slotchange);
+    }
+    
+    @JsxGetter({CHROME, EDGE, FF, FF_ESR})
+    public Object getFullscreenElement() {
+        return fullscreenElement_;
+    }
+    
+    @JsxGetter({CHROME, EDGE, FF, FF_ESR})
+    public Object getHost() {
+        return host_;
+    }
+    
+    @JsxGetter({CHROME, EDGE, FF, FF_ESR})
+    public String getInnerHTML() {
+        final StringBuilder buf = new StringBuilder();
+        for (DomNode child : getChildren().getElements()) {
+            if (child instanceof DomElement) {
```

Comment:
Pretty sure that only subclasses of **DomElement** should be represented in the assembled HTML fragment.